File: js/Neetcode_150/4_Binary_Search/24_min_in_rsa.js
//Find Min. In Rotated Sorted Srray: Given the sorted rotated array nums of unique elements, return the minimum element of this array.

var findMin = function(nums) {
    let ans = nums[0]
    let l = 0, r = nums.length-1
    while(l<=r){
        //if array is completely sorted, ans = nums[0]
        if(nums[l] <= nums[r]) return Math.min(ans, nums[l])

        let m = Math.floor(l + (r - l) / 2);
        ans = Math.min(ans, nums[m])
        // if we're in left sorted part, move to right sorted
        if(nums[l] <= nums[m]) l = m+1
        //if we're in right sorted part, move to left sorted part
        else r = m-1
    }
    return ans
};

File: c++/Neetcode_150/4_Binary_Search/24_min_in_rsa.cpp
//Find Min. In Rotated Sorted Srray: Given the sorted rotated array nums of unique elements, return the minimum element of this array.

int findMin(vector<int>& nums) {
        int s = 0, e = nums.size()-1;
        int ans = nums[0];
        while(s<=e){
            //if search space is already sorted
            if(nums[s] <= nums[e]){
                ans = min(ans, nums[s]);
                break;
            }
            int mid = s+ ((e-s)/2);
            //if left part[s->m] is sorted
            if(nums[s]<=nums[mid]){
                ans = min(ans, nums[s]);
                // now move to right part as min. of left part(ie, nums[s]) is already obtained
                s = mid+1;
            }
            //if right part[m->e] is sorted
            else{
                ans = min(ans, nums[mid]);
                // now move to leftt part as min. of rightt part(ie, nums[mid]) is already obtained
                e = mid-1;
            }
        }
        return ans;
    }